### PR TITLE
feat(cheatcodes): add resetGasMetering cheatcode

### LINKED ISSUE #4211: `vm.resetGasMetering()`: Reset gas metering
### Component

Forge

### Describe the feature you would like

We have `vm.resumeGasMetering()` and `vm.pauseGasMetering()`. While useful, the functionality I really end up needing is setting the gas metering to `0`, after I have completed my setup to in order to accurately compare the gas costs of two functions. This would also allow to factor out the gas costs of the function selector, which is a pain to zero-out manually.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
closes #4211 

## Solution

- group gas related state in `GasMetering` struct
- add `resetGasMeteringCall` / `vm.resetGasMetering()` 
- in interpreter step reset to gas limit if reset flag set to true

### CODE REVIEW

File: crates/forge/tests/cli/test_cmd.rs [hunk 1]
```diff
@@ -1481,3 +1481,114 @@ Traces:
 ...
 "#]]);
 });
+
+forgetest_init!(gas_metering_reset, |prj, cmd| {
+    prj.wipe_contracts();
+    prj.insert_ds_test();
+    prj.insert_vm();
+    prj.clear();
```
Comment (DaniPopes) [inline crates/forge/tests/cli/test_cmd.rs:1489]: what is this for? the project starts out with forgestd already

File: crates/cheatcodes/src/inspector.rs [hunk 2]
```diff
@@ -217,23 +217,42 @@ pub struct BroadcastableTransaction {
...
-            self.resumed = true;
+            self.touched = true;
         }
         self.paused_frames.clear();
     }
+
+    /// Reset gas to limit.
+    pub fn reset(&mut self) {
+        self.paused = false;
+        self.touched = true;
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:252]: does touched need to be set here too? can some combo of pause/resume/reset trigger out of gas without it?

File: crates/forge/tests/cli/test_cmd.rs (hunk 1 above)
Comment (grandizzy) [reply crates/forge/tests/cli/test_cmd.rs:1489]: correct, thing is that the cheatcode is not yet available in forge-std therefore used `insert_vm` to import local Vm

File: crates/cheatcodes/spec/src/vm.rs [hunk 3]
```diff
@@ -658,6 +658,10 @@ interface Vm {
     #[cheatcode(group = Evm, safety = Safe)]
     function pauseGasMetering() external;
 
+    /// Reset gas metering (i.e. gas usage is set to gas limit).
+    #[cheatcode(group = Evm, safety = Safe)]
+    function resetGasMetering() external;
```
[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/cheatcodes/spec/src/vm.rs]: moved in https://github.com/foundry-rs/foundry/pull/8750/commits/30311cc8d5ef06efed807e1cc6e07932f5ca717f

File: crates/cheatcodes/src/inspector.rs (hunk 2 above)
Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:252]: yeah, it is basically the same as the pause/resume scenario where we refund after reset, I added a test in https://github.com/foundry-rs/foundry/pull/8750/commits/30311cc8d5ef06efed807e1cc6e07932f5ca717f which would result in `[PASS] testResetNegativeGas() (gas: 18446744073709547520)` if touched not set

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:252]: this though raised an issue with gas reconciliation which was not consistent with gas reset (was setting to remaining gas instead limit), fixed in https://github.com/foundry-rs/foundry/pull/8750/commits/be96028928a5a0e04b513c7597206fdd112b351e gas properly showing 0 now if spent < refunded

File: crates/cheatcodes/src/inspector.rs [hunk 4]
```diff
@@ -264,12 +283,6 @@ pub struct Cheatcodes {
     /// execution block environment.
     pub block: Option<BlockEnv>,
 
-    /// The gas price
-    ///
-    /// Used in the cheatcode handler to overwrite the gas price separately from the gas price
-    /// in the execution environment.
-    pub gas_price: Option<U256>,
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:271]: idk if this makes sense to be moved, this is just used in initialize_interpreter to set env

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs]: makes sense, moved back in https://github.com/foundry-rs/foundry/pull/8750/commits/27f02ecf6a6e429f3fc3a7b2daf77eaf3da096a7

**Review by DaniPopes — APPROVED** [review]
lgtm

**Pull request merged**